src/topology/order/scott.lean

```diff
@@ -0,0 +1,469 @@
+  is_lub d a → is_lub (f '' d) (f(a))
+
+lemma preserve_lub_on_directed_montotone (f : α → β) (h: preserve_lub_on_directed f): monotone f :=
+begin
+  intros a b hab,
+  rw preserve_lub_on_directed at h,
+  let d := ({a, b} : set α),
+  have e1: is_lub (f '' d) (f b) :=
+  begin
+    apply h,
+    { exact insert_nonempty a {b} },
+    { exact directed_on_pair a b hab },
+    { rw is_lub,
+      split,
+      { simp only [upper_bounds_insert, upper_bounds_singleton, mem_inter_iff, mem_Ici, le_refl,
+          and_true],
+        exact hab, },
+      { simp only [upper_bounds_insert, upper_bounds_singleton],
+        rw (inter_eq_self_of_subset_right (Ici_subset_Ici.mpr hab)),
+        exact λ {x : α}, mem_Ici.mpr, } },
+  end,
+  rw [is_lub, is_least] at e1,
+  cases e1,
+  apply e1_left,
+  rw mem_image,
+  use a,
+  simp only [mem_insert_iff, eq_self_iff_true, true_or, and_self],
+end
+
+end preorder
+
+/--
+Type synonym for a preorder equipped with the Scott topology
+-/
+def with_scott_topology := α
+
+variables {α β}
+
+namespace with_scott_topology
+
+/-- `to_scott` is the identity function to the `with_scott_topology` of a type.  -/
+@[pattern] def to_scott : α ≃ with_scott_topology α := equiv.refl _
+
+/-- `of_scott` is the identity function from the `with_scott_topology` of a type.  -/
+@[pattern] def of_scott : with_scott_topology α ≃ α := equiv.refl _
+
+@[simp] lemma to_scott_symm_eq : (@to_scott α).symm = of_scott := rfl
+@[simp] lemma of_scott_symm_eq : (@of_scott α).symm = to_scott := rfl
+@[simp] lemma to_scott_of_scott (a : with_scott_topology α) : to_scott (of_scott a) = a := rfl
+@[simp] lemma of_scott_to_scott (a : α) : of_scott (to_scott a) = a := rfl
+@[simp] lemma to_scott_inj {a b : α} : to_scott a = to_scott b ↔ a = b := iff.rfl
+@[simp] lemma of_scott_inj {a b : with_scott_topology α} : of_scott a = of_scott b ↔ a = b :=
+iff.rfl
+
+/-- A recursor for `with_scott_topology`. Use as `induction x using with_scott_topology.rec`. -/
+protected def rec {β : with_scott_topology α → Sort*}
+  (h : Π a, β (to_scott a)) : Π a, β a := λ a, h (of_scott a)
+
+
+instance [nonempty α] : nonempty (with_scott_topology α) := ‹nonempty α›
+instance [inhabited α] : inhabited (with_scott_topology α) := ‹inhabited α›
+
+end with_scott_topology
+
+section preorder
+
+variables [preorder α] [preorder β]
+
+instance : preorder (with_scott_topology α) := ‹preorder α›
+
+instance : topological_space (with_scott_topology α) :=
+  (upper_set_topology ⊔ directed_lub_mem_implies_tail_subset)
+
+namespace with_scott_topology
+
+lemma is_open_eq_upper_and_lub_mem_implies_tail_subset (u : set (with_scott_topology α)) : is_open u
+= (is_upper_set u ∧ ∀ (d : set α) (a : α), d.nonempty → directed_on (≤) d → is_lub d a → a ∈ u
+  → ∃ b ∈ d, (Ici b) ∩ d ⊆ u) := rfl
+
+lemma is_open_eq_upper_and_lub_mem_implies_inter_nonempty (u : set (with_scott_topology α)) :
+is_open u = (is_upper_set u ∧
```

Comment:
This should be stated as an `iff` not an equality.

Reply:
Done, thanks.